include/zigbee/zigbee_app_utils.rst

```diff
@@ -277,29 +282,49 @@ It is used in :c:func:`zigbee_default_signal_handler` by default.
 
-  * For the end device, the application alarm is scheduled with ``stop_network_rejoin(ZB_TRUE)``, to be called after the amount of time specified in ``ZB_DEV_REJOIN_TIMEOUT_MS``.
-    Once called, the alarm stops the rejoin.
+  * Additionally for the end device, if :c:func:`stop_network_rejoin` is called with ``was_scheduled`` set to ``ZB_TRUE``, :c:func:`user_input_indicate` can restart the rejoin procedure.
+    :c:func:`user_input_indicate` restarts the rejoin procedure if the device did not join the network and is not trying to join a network.
+    It is safe to call this function from an interrupt and to call it multiple times.
 
-* When ``stop_network_rejoin(was_scheduled)`` is called, the network rejoin is canceled and the alarms scheduled by :c:func:`start_network_rejoin` are canceled.
+.. figure:: /images/zigbee_signal_handler_10_rejoin_router.svg
+   :alt: Rejoining network procedure for routers
 
-  * Additionally for the end device, if :c:func:`stop_network_rejoin` is called with ``was_scheduled`` set to ``ZB_TRUE``, :c:func:`user_input_indicate` can restart the rejoin procedure.
+   Rejoining network procedure for routers
+
+End device rejoining network
+++++++++++++++++++++++++++++
+
+The end device uses the default signal handler to try to join or rejoin the network for a finite period of time, because the end devices are often powered by batteries.
+
+* The procedure to join or rejoin the network is restarted after the device reset or power cycle.
+* The procedure to join or rejoin the network can be restarted by calling :c:func:`user_input_indicate`, but it needs to be implemented in the application (for example, by calling :c:func:`user_input_indicate` when a button is pressed).
+  The procedure is restarted only if the device does not join and the procedure is not running.
+
+When :c:func:`start_network_rejoin` is called, the rejoin procedure is started.
+For the end device, the application alarm is scheduled with ``stop_network_rejoin(ZB_TRUE)``, to be called after the amount of time specified in ``ZB_DEV_REJOIN_TIMEOUT_MS``.
+Once called, the alarm stops the rejoin.
+
+When ``stop_network_rejoin(was_scheduled)`` is called, the network rejoin is canceled and the alarms scheduled by :c:func:`start_network_rejoin` are canceled.
+If :c:func:`stop_network_rejoin` is called with ``was_scheduled`` set to ``ZB_TRUE``, :c:func:`user_input_indicate` can restart the rejoin procedure.
+:c:func:`user_input_indicate` restarts the rejoin procedure if the device did not join the network and is not trying to join a network.
+It is safe to call this function from an interrupt and to call it multiple times.
 
-* For end devices only, :c:func:`user_input_indicate` restarts the rejoin procedure if the device did not join the network and is not trying to join a network.
-  It is safe to call this function from an interrupt and to call it multiple times.
+.. figure:: /images/zigbee_signal_handler_10_rejoin_end_device.svg
```

Comment:
I think this figure needs to be improved: the `User action` is not part of a regular flow, meaning you do not check for user action after the whole rejoin procedure stops. This should be rather presented as an asynchronous event.